### PR TITLE
Torchscript conversions

### PR DESCRIPTION
**Pull request by tusharsangam**
Converting all Pytorch based policies in torchscript.
We will use torchscript to load new & old policies.

Why is it useful ?
we don't have to import habitat-lab library since models are loaded in torch script's intermediate representation thus avoiding Pytorch class imports to instantiate.  Plug & play.

How to use it ?
Download new weights.zip (will be shared in slack) & replace entire weights folder located at "spot_rl_experiments/"

Backward Compatibility:
You don't have to always use torchscript weights, you can switch to .pth (normal Pytorch) weights in config.yaml & code will take care of loading them correctly.

### CODE REVIEW

File: spot_rl_experiments/configs/config.yaml [hunk 1]
```diff
@@ -1,16 +1,15 @@
 WEIGHTS:
   # Regular Nav
-  NAV: "weights/CUTOUT_WT_True_SD_200_ckpt.99.pvp.pth"
-
+  NAV: "weights/torchscript/CUTOUT_WT_True_SD_200_ckpt.99.pvp_combined_net.torchscript" #"weights/CUTOUT_WT_True_SD_200_ckpt.99.pvp.pth"
```
Comment (jimmytyyang) [inline spot_rl_experiments/configs/config.yaml:3]: Could you please clean up the "#" sign?

File: spot_rl_experiments/utils/_deletelater_backup_pytorch_to_torchscript.py [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+import sys
+from collections import OrderedDict
+from typing import Any
+
+import numpy as np
+import torch
+from gym import spaces
+from gym.spaces import Dict as SpaceDict
+
+# from habitat_baselines.rl.ddppo.policy.resnet_policy import PointNavResNetPolicy
```
Comment (jimmytyyang) [inline spot_rl_experiments/utils/_deletelater_backup_pytorch_to_torchscript.py:18]: do you want to remove this?

File: spot_rl_experiments/configs/config.yaml [hunk 3]
```diff
@@ -1,16 +1,15 @@
...
 
   # Mobile Gaze torchscript module files path
-  MOBILE_GAZE_NET: "weights/mobile_gaze/mg97_2_latest_net.pytorch"
-  MOBILE_GAZE_ACTION_DIS: "weights/mobile_gaze/mg97_2_latest_action.pytorch"
-  MOBILE_GAZE_STD: "weights/mobile_gaze/mg97_2_latest_std.pth"
+  MOBILE_GAZE: "weights/torchscript/mg97_2_latest_combined_net.torchscript"
+  
   # Static place
-  PLACE: "weights/final_paper/place_10deg_32_seed300_1649709235_ckpt.75.pth"
+  PLACE: "weights/torchscript/place_10deg_32_seed300_1649709235_ckpt.75_combined_net.torchscript" #"weights/final_paper/place_10deg_32_seed300_1649709235_ckpt.75.pth"
```
Comment (KavitShah1998) [inline spot_rl_experiments/configs/config.yaml:12]: Nice!!

File: spot_rl_experiments/utils/_deletelater_backup_pytorch_to_torchscript.py [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+            PolicyActionData = self.policy.act(
+                batch,
+                self.test_recurrent_hidden_states,
+                self.prev_actions,
+                self.not_done_masks,
+                deterministic=True,
+            )
+            actions = PolicyActionData.actions
+            self.test_recurrent_hidden_states = PolicyActionData.rnn_hidden_states
+            # print("original hidden state:", PolicyActionData.rnn_hidden_states)
```
Comment (jimmytyyang) [inline spot_rl_experiments/utils/_deletelater_backup_pytorch_to_torchscript.py:176]: remove this please!

File: spot_rl_experiments/utils/_deletelater_backup_pytorch_to_torchscript.py [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+        else:
+            low, high = np.finfo(np.float32).min, np.finfo(np.float32).max
+        observation_space[key] = spaces.Box(
+            low=low,
+            high=high,
+            shape=shape,
+            dtype=np.float32,
+        )
+        observations[key] = np.zeros(shape, dtype=np.float32)
+    # print({key:value.shape for key, value in observations.items()})
```
Comment (jimmytyyang) [inline spot_rl_experiments/utils/_deletelater_backup_pytorch_to_torchscript.py:267]: remove this

File: spot_rl_experiments/utils/pytorch_to_torchscript.py [hunk 6]
```diff
@@ -285,7 +392,7 @@ def convert_from_pytorch_to_torchscript(conversion_params_yaml_path: str):
     recurrent_hidden_state_pytorch = pytorchpolicyconverter.test_recurrent_hidden_states
 
     print("actions from pytorch model:", actions_pytorch)
-    print("recurrent_hidden_state from pytorch model:", recurrent_hidden_state_pytorch)
+    # print("recurrent_hidden_state from pytorch model:", recurrent_hidden_state_pytorch)
```
Comment (jimmytyyang) [inline spot_rl_experiments/utils/pytorch_to_torchscript.py:395]: remove this

**Review by jimmytyyang — COMMENTED** [review]
Hi Tushar, a small request! Could you please add a small paragraph in the readme for how to do torch conversion? That will be great to have this!

File: spot_rl_experiments/spot_rl/real_policy.py [hunk 7]
```diff
@@ -3,6 +3,8 @@
 # LICENSE file in the root directory of this source tree.
 
 
+import math
```
Comment (KavitShah1998) [inline spot_rl_experiments/spot_rl/real_policy.py:6]: Seems like math is not getting used in this file. 
Can we remove it in that case?

File: spot_rl_experiments/utils/cparamsgaze.yaml [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+  joint: [[4,], '0.0', '1.0', 'np.float32']
+  is_holding: [[1,], '0.0', '1.0', 'np.float32']
+# Define the action space output length
+ACTION_SPACE_LENGTH: 4
+# The path to load and save the models
+TARGET_HAB3_POLICY_PATH: "../weights/pth/final_paper/gaze_normal_32_seed100_1649708902_ckpt.38.pth"
+OUTPUT_COMBINED_NET_SAVE_PATH: "../weights/torchscript/gaze_normal_32_seed100_1649708902_ckpt.38_combined_net.torchscript"
+# If we want to use stereo pair camera for mobile gaze
+USE_STEREO_PAIR_CAMERA: False
+NEW_HABITAT_LAB_POLICY_OR_OLD: 'old'
```
Comment (KavitShah1998) [inline spot_rl_experiments/utils/cparamsgaze.yaml:16]: Add a new line at EOF (end of file)

File: spot_rl_experiments/utils/cparamsmgaze.yaml [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+  spot_head_stereo_depth_sensor: [[240, 228, 1], '0.0', '1.0', 'np.float32']
+  joint: [[4,], 'np.finfo(np.float32).min', 'np.finfo(np.float32).max', 'np.float32']
+# Define the action space output length
+ACTION_SPACE_LENGTH: 7
+# The path to load and save the models
+TARGET_HAB3_POLICY_PATH: "../weights/pth/mg97_2_latest.pth"
+OUTPUT_COMBINED_NET_SAVE_PATH: "../weights/torchscript/mg97_2_latest_combined_net.torchscript"
+# If we want to use stereo pair camera for mobile gaze
+USE_STEREO_PAIR_CAMERA: False
+NEW_HABITAT_LAB_POLICY_OR_OLD: 'new'
```
Comment (KavitShah1998) [inline spot_rl_experiments/utils/cparamsmgaze.yaml:16]: Add a new line at EOF

File: spot_rl_experiments/utils/cparamsnav.yaml [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+  goal_heading: [[1,], '-np.pi', 'np.pi', 'np.float32']
+  target_point_goal_gps_and_compass_sensor: [[2,], 'np.finfo(np.float32).min', 'np.finfo(np.float32).max', 'np.float32']
+# Define the action space output length
+ACTION_SPACE_LENGTH: 2
+# The path to load and save the models
+TARGET_HAB3_POLICY_PATH: "../weights/pth/CUTOUT_WT_True_SD_200_ckpt.99.pvp.pth"
+OUTPUT_COMBINED_NET_SAVE_PATH: "../weights/torchscript/CUTOUT_WT_True_SD_200_ckpt.99.pvp_combined_net.torchscript"
+# If we want to use stereo pair camera for mobile gaze
+USE_STEREO_PAIR_CAMERA: False
+NEW_HABITAT_LAB_POLICY_OR_OLD: 'old'
```
Comment (KavitShah1998) [inline spot_rl_experiments/utils/cparamsnav.yaml:16]: Add a new line at EOF

File: spot_rl_experiments/utils/cparamsplace.yaml [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+  joint: [[4,], '0.0', '1.0', 'np.float32']
+  obj_start_sensor: [[3,], '0.0', '1.0', 'np.float32']
+# Define the action space output length
+ACTION_SPACE_LENGTH: 4
+# The path to load and save the models
+TARGET_HAB3_POLICY_PATH: "../weights/pth/final_paper/place_10deg_32_seed300_1649709235_ckpt.75.pth"
+OUTPUT_COMBINED_NET_SAVE_PATH: "../weights/torchscript/place_10deg_32_seed300_1649709235_ckpt.75_combined_net.torchscript"
+# If we want to use stereo pair camera for mobile gaze
+USE_STEREO_PAIR_CAMERA: False
+NEW_HABITAT_LAB_POLICY_OR_OLD: 'old'
```
Comment (KavitShah1998) [inline spot_rl_experiments/utils/cparamsplace.yaml:14]: Add a new line at EOF

File: spot_rl_experiments/configs/config.yaml (hunk 1 above)
Comment (tusharsangam) [reply spot_rl_experiments/configs/config.yaml:3]: That comment was kept if anyone wants to use previous pytorch weights, they can just uncomment & use pytorch weights without any changes in the code.
If you want to me to remove these comments, I can remove those.

File: spot_rl_experiments/utils/_deletelater_backup_pytorch_to_torchscript.py (hunk 2 above)
Comment (tusharsangam) [reply spot_rl_experiments/utils/_deletelater_backup_pytorch_to_torchscript.py:18]: Deleted this file

File: spot_rl_experiments/spot_rl/real_policy.py (hunk 7 above)
Comment (tusharsangam) [reply spot_rl_experiments/spot_rl/real_policy.py:6]: removed math

**Review by jimmytyyang — APPROVED** [review]
LGTM!

**Review by KavitShah1998 — APPROVED** [review]

**Pull request merged**